File: examples/greeting-service/src/main/java/greeting/example/GreetingController.java
package greeting.example;

import io.swagger.v3.oas.annotations.Hidden;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.bind.annotation.*;

import javax.validation.constraints.Pattern;

@RestController
public class GreetingController {


    private final GreetingService greetingService;

    public GreetingController(GreetingService greetingService) {
        this.greetingService = greetingService;
    }

    @RequestMapping("/greeting")
    public Greeting greeting(@RequestParam(value="name", defaultValue="World") @Pattern(regexp = "\\D+") String name) {
        return greetingService.greeting(name);
    }

    @PostMapping("/greeting")
    public Greeting greetingByPost(@RequestBody Greeting greeting) {
        return greetingService.greeting(greeting.getContent());
    }

    @DeleteMapping("/greeting")
    @Hidden
    public ResponseEntity<?> deleteGreeting() {
        MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
        headers.add("Foo", "Bar");
        return new ResponseEntity<>(headers, HttpStatus.NO_CONTENT);
    }


    @RequestMapping("/greeting-status")
    @ResponseStatus(code = HttpStatus.CREATED)
    public Greeting greetingWithStatus(@RequestParam(value="name", defaultValue="World") @Pattern(regexp = "\\D+") String name) {
        return greetingService.greeting(name);
    }
}

File: examples/greeting-service-kotlin/src/main/kotlin/greeting/example/GreetingController.kt
package greeting.example

import io.swagger.v3.oas.annotations.Hidden
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.ui.Model
import org.springframework.util.LinkedMultiValueMap
import org.springframework.util.MultiValueMap
import org.springframework.web.bind.annotation.*

import javax.validation.constraints.Pattern

@RestController
class GreetingController(val greetingService: GreetingService) {

    @GetMapping("/greeting")
    fun greeting(@RequestParam(value = "name", defaultValue = "World") @Pattern(regexp = "\\D+") name: String): Greeting {
        return greetingService.greeting(name)
    }

    @PostMapping("/greeting")
    fun greetingByPost(@RequestBody greeting: Greeting): Greeting {
        return greetingService.greeting(greeting.content)
    }

    @DeleteMapping("/greeting")
    @Hidden
    fun deleteGreeting(): ResponseEntity<Any> {
        val headers: MultiValueMap<String, String> = LinkedMultiValueMap()
        headers.add("Foo", "Bar")
        return ResponseEntity(headers, HttpStatus.NO_CONTENT)
    }


    @GetMapping("/greeting-status")
    @ResponseStatus(code = HttpStatus.CREATED)
    fun greetingWithStatus(
            @RequestParam(value = "name", defaultValue = "World")
            @Pattern(regexp = "\\D+") name: String): Greeting {
        return greetingService.greeting(name)
    }

    @GetMapping(path = ["/"], produces = ["text/html"])
    fun home(model: Model): String {
        model.addAttribute(
                "message",
                "Welcome to Micronaut for Spring")
        return "home"
    }
}